### ISSUE TITLE
Compiler error with `or` between struct and string type

### ISSUE THREAD

**Issue by rcorre** [vlang/v#10174]
**V doctor:**
```
OS: linux, "Arch Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 5600X 6-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /tmp
vmodules: /home/rcorre/.vmodules
vroot: /home/rcorre/src/v
vexe: /home/rcorre/src/v/v
vexe mtime: 2021-05-22 20:11:22
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 f1174da.851d9e1

Git version: git version 2.31.1
Git vroot status: weekly.2021.20-45-g851d9e14
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg example.v`
```v
import json

struct Request {}

fn parse(s string) ?Request {
	return json.decode(Request, s)
}

fn parse2(s string) ?Request {
	req := json.decode(Request, s) ?
	return req
}

fn main() {
	println( parse('{"a": ') or { 'error' })
	println( parse2('{"a": ') or { 'error' })
}

```

**What did you expect to see?**

no error

**What did you see instead?**
```
==================
/tmp/v/example.14519596016699552345.tmp.c:9442: error: cannot convert 'struct string' to 'struct main__Request'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

**Comment by danieldaeschle** [comment]
there should be a C error.

The correct code would be
```v
println( parse('{"a": ') or { Request{} })
// or
println( parse('{"a": ') or { panic(err) })
```

**Comment by rcorre** [comment]
Ah, I was under the impression C errors should never occur, and the compiler should reject this at the "V"-level (specifically due to "C error. This should never happen.")

**Comment by danieldaeschle** [comment]
yes, this needs definitely to be fixed. But the code was still wrong. A V error is expected.

**Comment by Dialga** [comment]
Fixed by #10354 [3f00ff4](https://github.com/vlang/v/commit/3f00ff465b3e3d6a8e5f4aedbff4b99344ff75a9).
